dan_layer/core/src/workers/pacemaker_worker.rs

```diff
@@ -0,0 +1,270 @@
+
+use super::hotstuff_error::HotStuffError;
+
+const LOG_TARGET: &str = "tari::dan_layer::pacemaker_worker";
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PacemakerWaitStatus {
+    WaitTimeOut,
+}
+
+async fn send_timeout_message<T: Debug + Send>(
+    wait_over: T,
+    tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+) -> Result<(), HotStuffError> {
+    info!(target: LOG_TARGET, "Sending timeout message for value: {:?}", wait_over);
+    tx_waiter_status
+        .send((wait_over, PacemakerWaitStatus::WaitTimeOut))
+        .await
+        .map_err(|_| HotStuffError::SendError)?;
+    Ok(())
+}
+
+/// A pacemaker service that is responsible for:
+///   1. Receiving [`start_wait`] messages parametrized over a type `T` and start wait up to a max_timeout value.
+///   2. Receiving [`stop_wait`] messages for an already existing [`start_wait`].
+///   3. If it receives a stop message then it should stop the waiting and do nothing else;
+///   4. Otherwise, it should notify the sender of [`start_wait`] message that the max_timeout duration has passed.
+#[derive(Debug)]
+pub struct Pacemaker<T> {
+    rx_start_signal: Receiver<T>,
+    rx_shutdown_signal: Receiver<T>,
+    tx_inner_map: HashMap<T, Sender<bool>>,
+    tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+    max_timeout: u64,
+}
+
+impl<T: Copy + Debug + PartialEq + Eq + Hash + Send + Sync + 'static> Pacemaker<T> {
+    pub fn spawn(
+        rx_start_signal: Receiver<T>,
+        rx_shutdown_signal: Receiver<T>,
+        tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+        max_timeout: u64,
+        shutdown: ShutdownSignal,
+    ) -> JoinHandle<Result<(), HotStuffError>> {
+        let pacemaker = Self::new(rx_start_signal, rx_shutdown_signal, tx_waiter_status, max_timeout);
+        tokio::spawn(pacemaker.run(shutdown))
+    }
+
+    fn new(
+        rx_start_signal: Receiver<T>,
+        rx_shutdown_signal: Receiver<T>,
+        tx_waiter_status: Sender<(T, PacemakerWaitStatus)>,
+        max_timeout: u64,
+    ) -> Self {
+        Self {
+            rx_start_signal,
+            rx_shutdown_signal,
+            tx_inner_map: HashMap::new(),
+            tx_waiter_status,
+            max_timeout,
+        }
+    }
+
+    pub async fn run(mut self, mut shutdown: ShutdownSignal) -> Result<(), HotStuffError> {
+        let max_timeout = self.max_timeout;
+        loop {
+            tokio::select! {
+                msg = self.rx_start_signal.recv() => {
+                    if let Some(wait_over) = msg {
+                        let (tx, mut rx) = channel::<bool>(1);
+                        self.tx_inner_map.insert(wait_over, tx);
+                        info!(
+                            target: LOG_TARGET,
+                            "Received start wait signal for value: {:?}", wait_over
+                        );
+                        let tx_waiter_status = self.tx_waiter_status.clone();
+                        let _join = tokio::spawn(async move {
```

Comment:
I see that you switch from the `FuturesUnordered` approched to this. Was it not feasible to do it without the tokio::spawns? I'm wondering what's the limit of the tokio::spawn. I remember when there are too much thread you will not be notified at all, it will just not run. So you will never receive a trigger. But maybe the number is high. I'm just curious.

Reply:
So I've tested this, and tokio doesn't spawn real threads (my test spawned 10k threads) and there were only 20 system threads. I don't know how tokio does it, but it's good.